anagrafica/tests.py

```diff
@@ -20,13 +20,216 @@
 from base.utils import poco_fa
 from base.utils_tests import crea_persona_sede_appartenenza, crea_persona, crea_sede, crea_appartenenza, email_fittizzia, \
     crea_utenza
-from formazione.models import Aspirante
+from formazione.models import Aspirante, PartecipazioneCorsoBase, CorsoBase
 from jorvik.settings import GOOGLE_KEY
 from posta.models import Messaggio, Autorizzazione
 
 
 class TestAnagrafica(TestCase):
 
+    def test_appartenenza_modificabile_con_campo_precedente(self):
+
+        persona = crea_persona()
+        sede1 = crea_sede()
+        sede2 = crea_sede()
+
+        # appartenenza "isolata"
+        a_persona_1 = crea_appartenenza(persona, sede1)
+        self.assertTrue(a_persona_1.modificabile())
+
+        # doppia appartenenza "aperta" non genera blocchi (anche se a rigore non consistente)
+        a_persona_2 = crea_appartenenza(persona, sede2)
+        a_persona_2.inizio = "1980-11-10"
+        a_persona_1.precedente = a_persona_2
+        a_persona_1.save()
+        a_persona_2.save()
+        self.assertTrue(a_persona_1.modificabile())
+
+        # Se l'appartenenza è terminata, quella più recente non è modificabile
+        a_persona_2.terminazione = Appartenenza.TRASFERIMENTO
+        a_persona_2.save()
+        a_persona_1.refresh_from_db()
+        a_persona_2.refresh_from_db()
+        self.assertFalse(a_persona_1.modificabile())
+
+        # Se la precedente è un'estensione non si può modificare
+        a_persona_2.terminazione = Appartenenza.FINE_ESTENSIONE
+        a_persona_2.membro = Appartenenza.ESTESO
+        a_persona_2.save()
+        a_persona_1.refresh_from_db()
+        a_persona_2.refresh_from_db()
+        self.assertFalse(a_persona_1.modificabile())
+
+        # Se è un trasferimento o altro non si può modificare
+        for tipo in Appartenenza.TERMINAZIONE:
+            if tipo[0] not in Appartenenza.MODIFICABILE_SE_TERMINAZIONI_PRECEDENTI:
+                for membro in Appartenenza.MEMBRO:
+                    if membro[0] not in Appartenenza.PRECEDENZE_MODIFICABILI:
+                        a_persona_2.terminazione = tipo[0]
+                        a_persona_2.membro = membro[0]
+                        a_persona_2.save()
+                        a_persona_1.refresh_from_db()
+                        a_persona_2.refresh_from_db()
+                        self.assertFalse(a_persona_1.modificabile())
+
+    def test_appartenenza_modificabile_con_precedente_per_data(self):
+
+        persona = crea_persona()
+        sede1 = crea_sede()
+        sede2 = crea_sede()
+
+        # appartenenza "isolata"
+        a_persona_1 = crea_appartenenza(persona, sede1)
+        self.assertTrue(a_persona_1.modificabile())
+
+        corso = CorsoBase.objects.create(
+            sede=sede1, data_inizio=poco_fa(), data_esame=poco_fa(), progressivo=1, anno=poco_fa().year
+        )
+        # Corso non completato non genera blocchi
```

Comment:
Questo modella il caso in cui ci sia già un'appartenenza inserita manualmente come volontario e una partecipazione ad un corso base aperto per qualcunque motivo